src/ILLink.Shared/XmlAttributeParser.cs

```diff
@@ -0,0 +1,354 @@
+			protected const string AssemblyElementName = "assembly";
+			protected const string ArgumentElementName = "argument";
+			protected const string ParameterElementName = "parameter";
+			protected const string ReturnElementName = "return";
+			protected const string AllAssembliesFullName = "*";
+			protected const string XmlNamespace = "";
+
+			public static List<IRootNode> ProcessRootNodes (XPathNavigator nav)
+			{
+				if (!nav.MoveToChild (LinkerElementName, XmlNamespace)) {
+					throw new ArgumentException ("XML does not have <linker> base tag");
+				}
+				var roots = new List<IRootNode> ();
+				foreach (XPathNavigator typeNav in nav.SelectChildren (TypeElementName, "")) {
+					roots.Add (new TypeNode (typeNav));
+				}
+				foreach (XPathNavigator assemblyNav in nav.SelectChildren (AssemblyElementName, "")) {
+					roots.Add (new TypeNode (assemblyNav));
+				}
+				return roots;
+			}
+			protected static string GetFullName (XPathNavigator nav)
+			{
+				return GetAttribute (nav, FullNameAttributeName);
+			}
+
+			protected static string GetName (XPathNavigator nav)
+			{
+				var name = GetAttribute (nav, NameAttributeName);
+				return name;
+			}
+
+			protected static string GetSignature (XPathNavigator nav)
+			{
+				return GetAttribute (nav, SignatureAttributeName);
+			}
+
+			protected static string GetAttribute (XPathNavigator nav, string attribute)
+			{
+				return nav.GetAttribute (attribute, XmlNamespace);
+			}
+		}
+
+		public abstract record AttributeTargetNode : NodeBase
+		{
+			public List<AttributeNode> Attributes;
+
+			public AttributeTargetNode (XPathNavigator nav) : base (nav)
+			{
+				var attributes = new List<AttributeNode> ();
+				foreach (XPathNavigator attributeNav in nav.SelectChildren (AttributeElementName, "")) {
+					var attr = new AttributeNode (attributeNav);
+					if (attr == null)
+						continue;
+					attributes.Add (attr);
+				}
+				Attributes = attributes;
+			}
+		}
+
+		public abstract record FeatureSwitchedNode : AttributeTargetNode
+		{
+			public FeatureSwitch? FeatureSwitch;
+
+			public FeatureSwitchedNode (XPathNavigator nav) : base (nav)
+			{
+				string? feature = GetAttribute (nav, "feature");
+				if (String.IsNullOrEmpty (feature))
+					feature = null;
+				string? featurevalue = GetAttribute (nav, "featurevalue");
+				if (String.IsNullOrEmpty (featurevalue))
+					featurevalue = null;
+				string? featuredefault = GetAttribute (nav, "featuredefault");
+				if (String.IsNullOrEmpty (featuredefault))
+					featuredefault = null;
+				if (feature is null && featurevalue is null && featuredefault is null) {
+					return;
+				}
+				// if (feature is null && (featurevalue is not null || featuredefault is not null)) ;
+				// error, feature must be defined if featurevalue or featuredefault is defined
+				FeatureSwitch = new FeatureSwitch (feature, featurevalue, featuredefault);
+			}
+		}
+
+		public record FeatureSwitch
```

Comment:
Won't it be over the recommended 16 bytes for a value type?

Reply:
Where is that recommended? I was just thinking that it has value semantics - it's just three pointers that we want to pass around together as a group. From a performance standpoint, I would guess that it's probably not worth  allocating an object on the heap to store those pointers, but I doubt this will be on the critical path for performance anyway.

Reply:
It's recommended here: https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/choosing-between-class-and-struct but since it's only really used when stored in other reference types and not passed/copied around the stack outside of constructors it probably wouldn't matter much.

Reply:
Thanks! I'll leave it up to you - I don't feel strongly about it.